src/main/java/eu/europeana/metis/sandbox/service/workflow/InternalValidationServiceImpl.java

```diff
@@ -11,54 +11,106 @@
-                    validationResult.getRecordId(), validationResult.getNodeId());
-        }
-        try {
-            generateAnalysis(recordToValidate.getDatasetId(), recordToValidate.getContent());
-        } catch (PatternAnalysisException e) {
-            LOGGER.error("An error occurred while processing pattern analysis with record id {}", recordToValidate.getEuropeanaId());
-        }
-        return new RecordInfo(recordToValidate);
+  private void generateAnalysis(String datasetId, byte[] recordContent) throws PatternAnalysisException {
+    final Lock lock = datasetIdLocksMap.computeIfAbsent(datasetId, s -> new ReentrantLock());
+    final ExecutionPoint executionPoint;
+    try {
+      lock.lock();
+      LOGGER.debug("{} lock, Locked", datasetId);
+      final LocalDateTime timestamp = datasetIdTimestampMap.computeIfAbsent(datasetId, s -> getLocalDateTime(datasetId));
+      //We have to attempt initialization everytime because we don't have an entry point for the start of the step
+      executionPoint = patternAnalysisService.initializePatternAnalysisExecution(datasetId, Step.VALIDATE_INTERNAL, timestamp);
+      // TODO: 03/05/2022 This needs to still be synchronized.
+      //  To fix this we'll need to do an upsert(with a spring jdbcTemplate perhaps) on the internal counter update in method insertPatternAnalysis
+      patternAnalysisService.generateRecordPatternAnalysis(executionPoint, new String(recordContent, StandardCharsets.UTF_8));
+    } finally {
+      lock.unlock();
+      LOGGER.debug("{} lock, Unlocked", datasetId);
     }
+  }
+
+  private LocalDateTime getLocalDateTime(String datasetId) {
+    //This only applies for sandbox, because the same id cannot have multiple executions.
+    //In metis for example this wouldn't work if there was already a valid execution of a step in the database
+    //while we are trying to run a new execution of the same step
+    Optional<ExecutionPoint> firstOccurrence = executionPointService.getExecutionPoint(datasetId,
+        Step.VALIDATE_INTERNAL.toString());
+    return firstOccurrence.map(ExecutionPoint::getExecutionTimestamp).orElseGet(LocalDateTime::now);
+  }
 
-    private void generateAnalysis(String datasetId, byte[] recordContent) throws PatternAnalysisException {
-        Optional<ExecutionPoint> firstOccurrence = executionPointService.getExecutionPoint(datasetId, Step.VALIDATE_INTERNAL.toString());
-        LocalDateTime timestamp = firstOccurrence.isPresent() ? firstOccurrence.get().getExecutionTimestamp() : LocalDateTime.now();
-        patternAnalysisService.generateRecordPatternAnalysis(datasetId, Step.VALIDATE_INTERNAL, timestamp,
-                new String(recordContent, StandardCharsets.UTF_8));
+  /**
+   * Evict cache items every day.
+   */
+  @Scheduled(cron = "0 0 0 * * ?")
+  private void cleanCache() {
+    for (Entry<String, Lock> entry : datasetIdLocksMap.entrySet()) {
+      final Lock lock = entry.getValue();
+      try {
+        lock.lock();
```

Comment:
It would be nice to have a ```LOGGER.info("Dataset map locks cached cleared");``` as a reminder that this task is running in the background an it is done.

Reply:
yes but I added it as debug not info.